File: BlackFireIM/Example/Lib/MSIMSDK-UIKit/ChatDetail/View/Message/MSVoiceMessageCell.m
//
//  MSVoiceMessageCell.m
//  BlackFireIM
//
//  Created by benny wang on 2021/4/22.
//

#import "MSVoiceMessageCell.h"
#import "UIView+Frame.h"


@implementation MSVoiceMessageCell

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        _voice = [[UIImageView alloc] init];
        _voice.animationDuration = 1;
        [self.bubbleView addSubview:_voice];

        _duration = [[UILabel alloc] init];
        _duration.font = [UIFont systemFontOfSize:12];
        _duration.textColor = [UIColor grayColor];
        [self.bubbleView addSubview:_duration];
    }
    return self;
}

- (void)dealloc
{
    NSLog(@"%@ dealloc",self.class);
    [self.voiceData removeObserver:self forKeyPath:@"isPlaying"];
}

- (void)fillWithData:(MSVoiceMessageCellData *)data
{
    //set data
    [super fillWithData:data];
    self.voiceData = data;
    if (data.voiceElem.duration > 0) {
        _duration.text = [NSString stringWithFormat:@"%ld\"", (long)data.voiceElem.duration];
    } else {
        _duration.text = @"1\"";    // 显示0秒容易产生误解
    }
    _voice.image = data.voiceImage;
    _voice.animationImages = data.voiceAnimationImages;
    
    //animate
    [self.voiceData addObserver:self forKeyPath:@"isPlaying" options:NSKeyValueObservingOptionNew context:nil];
    self.voiceData.isPlaying = self.voiceData.isPlaying;
    
    if (data.direction == MsgDirectionIncoming) {
        _duration.textAlignment = NSTextAlignmentLeft;
        _duration.textColor = [UIColor grayColor];
    } else {
        _duration.textAlignment = NSTextAlignmentRight;
        _duration.textColor = [UIColor whiteColor];
    }
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSString *,id> *)change context:(void *)context
{
    NSNumber *isPlaying = [change objectForKey:NSKeyValueChangeNewKey];
    if ([isPlaying boolValue]) {
        [self.voice startAnimating];
    }else {
        [self.voice stopAnimating];
    }
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    if (self.voiceData.direction == MsgDirectionIncoming) {
        self.voice.frame = CGRectMake(16, 20-self.voiceData.voiceImage.size.height*0.5, self.voiceData.voiceImage.size.width, self.voiceData.voiceImage.size.height);
        self.duration.frame = CGRectMake(self.voice.maxX+10, 0, 40, 20);
        self.duration.centerY = self.voice.centerY;
    } else {
        self.voice.frame = CGRectMake(0, 20-self.voiceData.voiceImage.size.height*0.5, self.voiceData.voiceImage.size.width, self.voiceData.voiceImage.size.height);
        self.voice.maxX = self.container.width-16;
        self.duration.frame = CGRectMake(0, 0, 40, 20);
        self.duration.centerY = self.voice.centerY;
        self.duration.maxX = self.voice.x-10;
    }
}

@end

File: Swift/MSIM-Swift/Lib/MSIMSDK-UIKit/ChatDetail/View/Message/MSVoiceMessageCell.swift
//
//  MSVoiceMessageCell.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/8/10.
//

import UIKit

open class MSVoiceMessageCell: MSBubbleMessageCell {

    public var voice: UIImageView!
    
    public var durationL: UILabel!
    
    public var voiceData: MSVoiceMessageCellData?
    
    var kvoToken: NSKeyValueObservation?
    
    public override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        
        voice = UIImageView()
        voice.animationDuration = 1
        bubbleView.addSubview(voice)
        
        durationL = UILabel()
        durationL.textColor = .gray
        durationL.font = .systemFont(ofSize: 12)
        bubbleView.addSubview(durationL)
    }
    
    required public init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    deinit {
        voiceData?.removeObserver(self, forKeyPath: "isPlaying")
    }
    
    public override func fillWithData(data: MSMessageCellData) {
        super.fillWithData(data: data)
        if let voiceData = data as? MSVoiceMessageCellData {
            self.voiceData = voiceData
            if voiceData.message.voiceElem!.duration > 0 {
                durationL.text = String(format: "%ld\"", voiceData.message.voiceElem!.duration)
            }else {
                durationL.text = "1\""
            }
            voice.image = voiceData.voiceImage
            voice.animationImages = voiceData.voiceAnimationImages
            
            self.observe(data: self.voiceData!)
            self.voiceData!.isPlaying = self.voiceData!.isPlaying
            
            if voiceData.direction == .inComing {
                durationL.textAlignment = .left
                durationL.textColor = .gray
            }else {
                durationL.textAlignment = .right
                durationL.textColor = .white
            }
        }
    }
    
    func observe(data: MSVoiceMessageCellData) {
        kvoToken = data.observe(\.isPlaying, options: .new, changeHandler: { voiceData, change in
            if let isPlaying = change.newValue {
                if isPlaying {
                    self.voice.startAnimating()
                }else {
                    self.voice.stopAnimating()
                }
            }
        })
    }
    
    open override func layoutSubviews() {
        super.layoutSubviews()
        if voiceData?.direction == .inComing {
            voice.frame = CGRect(x: 16, y: 20 - voiceData!.voiceImage.size.height * 0.5, width: voiceData!.voiceImage.size.width, height: voiceData!.voiceImage.size.height)
            durationL.frame = CGRect(x: voice.right + 10, y: 0, width: 40, height: 20)
            durationL.centerY = voice.centerY
        }else {
            voice.frame = CGRect(x: 0, y: 20 - voiceData!.voiceImage.size.height * 0.5, width: voiceData!.voiceImage.size.width, height: voiceData!.voiceImage.size.height)
            voice.right = container.width - 16
            durationL.frame = CGRect(x: 0, y: 0, width: 40, height: 20)
            durationL.centerY = voice.centerY
            durationL.right = voice.left - 10
        }
    }
}
